File: libs/java/core/src/main/java/ca/bc/gov/tno/models/ContentBase.java
package ca.bc.gov.tno.models;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

import ca.bc.gov.tno.ContentType;

/**
 * ContentBase class, provides a model for capturing all the relevant data that
 * will make content searchable.
 */
public abstract class ContentBase {
  /**
   * The source of the content, which can be a URL or identifier of the source
   * like their name.
   */
  private String source;

  /**
   * The type of content.
   */
  private ContentType type;

  /**
   * The unique identifier from the source that identifies this content
   * throughout.
   */
  private String uid;

  /**
   * URL to the original content.
   */
  private String link;

  /**
   * The language of the content.
   */
  private String language;

  /**
   * The source copyright.
   */
  private String copyright;

  /**
   * The name of the author.
   */
  private String author;

  /**
   * The content title.
   */
  private String title;

  /**
   * The content summary.
   */
  private String summary;

  /**
   * The path to the source content file (audio/video/image).
   */
  private String filePath;

  /**
   * The URL to the stream.
   */
  private String streamUrl;

  /**
   * When the content was published on.
   */
  private Date publishedOn;

  /**
   * When the content was last updated on.
   */
  private Date updatedOn;

  /**
   * A collection of tags that are used to search for this content.
   */
  private List<Tag> tags = new ArrayList<Tag>();

  /**
   * Creates a new instance of a ContentBase object.
   */
  public ContentBase() {
  }

  /**
   * Creates a new instance of a ContentBase object, initializes with specified
   * parameters.
   * 
   * @param source The unique identifier of the source (i.e. code).
   * @param type   The type of content [text|audio|video].
   * @param uid    A unique identifier for the content.
   * @param title  The content title.
   */
  public ContentBase(String source, ContentType type, String uid, String title) {
    if (source == null || source.length() == 0)
      throw new IllegalArgumentException("Parameter 'source' is required, and cannot be empty.");
    if (uid == null || uid.length() == 0)
      throw new IllegalArgumentException("Parameter 'uid' is required, and cannot be empty.");
    if (title == null || title.length() == 0)
      throw new IllegalArgumentException("Parameter 'title' is required, and cannot be empty.");

    this.source = source;
    this.type = type;
    this.uid = uid;
    this.title = title;
  }

  /**
   * @return String return the source
   */
  public String getSource() {
    return source;
  }

  /**
   * @param source the source to set
   */
  public void setSource(String source) {
    this.source = source;
  }

  /**
   * @return ContentType return the type
   */
  public ContentType getType() {
    return type;
  }

  /**
   * @param type the type to set
   */
  public void setType(ContentType type) {
    this.type = type;
  }

  /**
   * @return String return the uid
   */
  public String getUid() {
    return uid;
  }

  /**
   * @param uid the uid to set
   */
  public void setUid(String uid) {
    this.uid = uid;
  }

  /**
   * @return String return the title
   */
  public String getTitle() {
    return title;
  }

  /**
   * @param title the title to set
   */
  public void setTitle(String title) {
    this.title = title;
  }

  /**
   * @return String return the summary
   */
  public String getSummary() {
    return summary;
  }

  /**
   * @param summary the summary to set
   */
  public void setSummary(String summary) {
    this.summary = summary;
  }

  /**
   * @return Date return the publishedOn
   */
  public Date getPublishedOn() {
    return publishedOn;
  }

  /**
   * @param publishedOn the publishedOn to set
   */
  public void setPublishedOn(Date publishedOn) {
    this.publishedOn = publishedOn;
  }

  /**
   * @return Date return the updatedOn
   */
  public Date getUpdatedOn() {
    return updatedOn;
  }

  /**
   * @param updatedOn the updatedOn to set
   */
  public void setUpdatedOn(Date updatedOn) {
    this.updatedOn = updatedOn;
  }

  /**
   * @return List{Tag} return the tags
   */
  public List<Tag> getTags() {
    return tags;
  }

  /**
   * @param tags the tags to set
   */
  public void setTags(List<Tag> tags) {
    this.tags = tags;
  }

  /**
   * @return String return the link
   */
  public String getLink() {
    return link;
  }

  /**
   * @param link the link to set
   */
  public void setLink(String link) {
    this.link = link;
  }

  /**
   * @return String return the language
   */
  public String getLanguage() {
    return language;
  }

  /**
   * @param language the language to set
   */
  public void setLanguage(String language) {
    this.language = language;
  }

  /**
   * @return String return the author
   */
  public String getAuthor() {
    return author;
  }

  /**
   * @param author the author to set
   */
  public void setAuthor(String author) {
    this.author = author;
  }

  /**
   * @return String return the filePath
   */
  public String getFilePath() {
    return filePath;
  }

  /**
   * @param filePath the filePath to set
   */
  public void setFilePath(String filePath) {
    this.filePath = filePath;
  }

  /**
   * @return String return the streamUrl
   */
  public String getStreamUrl() {
    return streamUrl;
  }

  /**
   * @param streamUrl the streamUrl to set
   */
  public void setStreamUrl(String streamUrl) {
    this.streamUrl = streamUrl;
  }

  /**
   * @return String return the copyright
   */
  public String getCopyright() {
    return copyright;
  }

  /**
   * @param copyright the copyright to set
   */
  public void setCopyright(String copyright) {
    this.copyright = copyright;
  }

}

File: libs/net/kafka/Models/ContentBase.cs
using TNO.Entities;

namespace TNO.Kafka.Models;

/// <summary>
/// ContentBase abstract class, provides a model for content sent to Kafka.
/// TODO: Change name to ContentBaseModel for consistent naming convention
/// </summary>
public abstract class ContentBase
{
    #region Properties

    /// <summary>
    /// get/set - A unique key to identify this content.
    /// </summary>
    public string Uid { get; set; } = "";

    /// <summary>
    /// get/set - A unique code to identify the data source the content belong to.
    /// </summary>
    public string Source { get; set; } = "";

    /// <summary>
    /// get/set - The type of content and form to use.
    /// </summary>
    public ContentType ContentType { get; set; }

    /// <summary>
    /// get/set - Foreign key to the product the content will be assigned by default.
    /// </summary>
    public int ProductId { get; set; }

    /// <summary>
    /// get/set - A URL to the content.
    /// TODO: Change to a Uri type.
    /// </summary>
    public string Link { get; set; } = "";

    /// <summary>
    /// get/set - The language of the content.
    /// </summary>
    public string Language { get; set; } = "";

    /// <summary>
    /// get/set - The copyright information for the content.
    /// </summary>
    public string Copyright { get; set; } = "";

    /// <summary>
    /// get/set - The title or headline to identify the content subject.
    /// </summary>
    public string Title { get; set; } = "";

    /// <summary>
    /// get/set - The summary or abstract of the content.
    /// </summary>
    public string Summary { get; set; } = "";

    /// <summary>
    /// get/set - A data location is where the source was ingested and the files may reside.
    /// </summary>
    public string DataLocation { get; set; } = "";

    /// <summary>
    /// get/set - The path to the file associated with this content.
    /// This path will be relative to a data location that is available to the consuming service.
    /// Essentially the path is meaningless without the additional context, but is critical to fetch the file.
    /// </summary>
    public string FilePath { get; set; } = "";

    /// <summary>
    /// get/set - The section containing the content.
    /// </summary>
    public string Section { get; set; } = "";

    /// <summary>
    /// get/set - The page number containing the content.
    /// </summary>
    public string Page { get; set; } = "";

    /// <summary>
    /// get/set - A URL to a stream for this content.
    /// TODO: Change to a Uri type.
    /// </summary>
    public string StreamUrl { get; set; } = "";

    /// <summary>
    /// get/set - The date this content was published.
    /// </summary>
    public DateTime? PublishedOn { get; set; }

    /// <summary>
    /// get/set - The date when this content was updated.
    /// </summary>
    public DateTime? UpdatedOn { get; set; }

    /// <summary>
    /// get/set - An array of author information.
    /// </summary>
    public IEnumerable<Author> Authors { get; set; } = Array.Empty<Author>();

    /// <summary>
    /// get/set - An array of tags to identify this content.
    /// </summary>
    public IEnumerable<Tag> Tags { get; set; } = Array.Empty<Tag>();
    #endregion

    #region Constructors
    /// <summary>
    /// Creates a new instance of a ContentBase object.
    /// </summary>
    public ContentBase() { }

    /// <summary>
    /// Creates a new instance of a ContentBase object, initializes with specified parameters.
    /// </summary>
    /// <param name="dataLocation"></param>
    /// <param name="source"></param>
    /// <param name="contentType"></param>
    /// <param name="productId"></param>
    /// <param name="uid"></param>
    /// <param name="title"></param>
    /// <param name="summary"></param>
    /// <exception cref="ArgumentNullException"></exception>
    public ContentBase(string dataLocation, string source, ContentType contentType, int productId, string uid, string title, string summary)
    {
        this.DataLocation = dataLocation;
        this.ContentType = contentType;
        this.ProductId = productId;
        this.Source = source ?? throw new ArgumentNullException(nameof(source));
        this.Uid = uid ?? throw new ArgumentNullException(nameof(uid));
        this.Title = title ?? throw new ArgumentNullException(nameof(title));
        this.Summary = summary ?? throw new ArgumentNullException(nameof(summary));
    }

    /// <summary>
    /// Creates a new instance of a ContentBase object, initializes with specified parameters.
    /// </summary>
    /// <param name="dataLocation"></param>
    /// <param name="source"></param>
    /// <param name="contentType"></param>
    /// <param name="productId"></param>
    /// <param name="uid"></param>
    /// <param name="title"></param>
    /// <param name="summary"></param>
    /// <param name="publishedOn"></param>
    public ContentBase(string dataLocation, string source, ContentType contentType, int productId, string uid, string title, string summary, DateTime publishedOn)
        : this(dataLocation, source, contentType, productId, uid, title, summary)
    {
        this.PublishedOn = publishedOn;
    }
    #endregion
}
